processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultColumnSelectorFactoryMaker.java

```diff
@@ -160,12 +160,17 @@ public ColumnValueSelector makeColumnValueSelector(@Nonnull String columnName)
     @Override
     public ColumnCapabilities getColumnCapabilities(String column)
     {
-      return withColumnAccessor(column, columnAccessor ->
-          new ColumnCapabilitiesImpl()
+      return withColumnAccessor(column, columnAccessor -> {
+        if (columnAccessor == null) {
+          return ColumnCapabilitiesImpl.createDefault();
+        } else {
```

Comment:
Hrm, why was the `columnAccessor` null?  I'm a bit curious as I wonder if this fix shouldn't be done in a different place

Reply:
query is:
```
  SELECT
    '_'||dim1,
    MIN(cast(0 as double)) OVER ()
  FROM foo
```
 * so there is effectively no column inputs to the agg
 * [AggregatorUtil](https://github.com/apache/druid/blob/d25caaefa4282a6b61663e8184d3c0b4092f6f35/processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java#L335) is looking for the selector
      * since it already enforces and exclusive or of setting `fieldName` and `fieldExpression` ; I didn't seen an easy way around there
      * its actually a const column; so that's why it doesn't really matter what capabilities it has....

yes; it could: `withColumnAccessor` gives [no gurantees](https://github.com/apache/druid/blob/d25caaefa4282a6b61663e8184d3c0b4092f6f35/processing/src/main/java/org/apache/druid/query/rowsandcols/semantic/DefaultColumnSelectorFactoryMaker.java#L176) at all that the accessor will be non-null - if that shouldn't happen it must throw an exception signaling that something is wrong

note: I think `ColumnAccessor` should be a (possibly lazily inited) field - I don't understand the benefits what the `private`  `withColumnAccessor` is doing with a `cache` which has no retention policy...it looks unneccaily added complexity for no good reason

I was looking into this and apparently this may partially be a sideeffect of that:
* a [virtualcolumnregistry](https://github.com/apache/druid/blob/d25caaefa4282a6b61663e8184d3c0b4092f6f35/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java#L207) is created during windowing expression build
* this lands as the `virtualColumnRegistry` in `DruidQuery` later on
* however in [toWindowQuery()](https://github.com/apache/druid/blob/d25caaefa4282a6b61663e8184d3c0b4092f6f35/sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java#L1445) a native query is created which can't even care about that - as the `virtualColumnRegistry` is not passed to that call

now I think that:
* `withColumnAccessor` must throw an exception in case the column is `null`
* `DruidQuery#virtualColumnRegistry` could be handled possibly:
   * accept VC-s in `WindowOperatorQuery`
   * or somehow push it down into the below scan query

Reply:
The point of the cache for `withColumnAccessor` is to reuse the column object if it's getting referenced multiple times.  Specifically, various of our columns do on-the-fly decompression and end up checking out buffers in order to store the decompressed values, those are both (1) resources that need to be cleaned up and (2) decompression is CPU work we want to minimize.  But, if you have multiple aggregators all depending on the same column, they will each independently ask for the column, so we have to deduplicate those requests somewhere and this layer is doing that.

The retention policy on the cache is for the lifecycle of the object.  The object itself only exists for the lifecycle of a single query against the data, so it's a relatively short-lived item.

I think the handling of this in this PR is actually correct.  The bug came about because when this code asked the underlying `RowsAndColumns` about the column (that is a virtual column), the `RowsAndColumns` returned `null` instead of the column.  As it exists currently, it's relatively hard to make a `VirtualColumn` that runs on top of a RowsAndColumns (`VirtualColumn` is pretty tightly coupled with the `StorageAdapter` stuff), so it makes sense to create the column through forced materialization (as is being done by your change).

My question about why it got a `null` was partially because I was wondering if `null` should be passed to the function or just returned.  But, `withColumnAccessor` is private and looking at the other lambda implementations, they are all doing `null` checks too, so it's probably six-of-one, half-dozen of another.